### ISSUE TITLE
NLL: Review compile-fail migrated tests

### ISSUE THREAD

**Issue by davidtwco** [rust-lang/rust#53351]
As part of #53196, `compile-fail` tests were migrated to `ui` tests. In doing this, a handful of tests compile successfully in the NLL compare mode, but not in the normal AST compare mode. These tests were marked as "ignore on NLL compare mode" as there isn't an easy way to assert that a test compiles successfully in one compare mode and not in another. These tests are listed below and should be checked to make sure that it is correct that these tests ran successfully in NLL compare mode or if they have a bug:

(exact locations of tests have changed)

## borrowck/

- [x] ui/compile-fail-migration/borrow-tuple-fields.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrow-tuple-fields.rs#L11 
- [x] ui/compile-fail-migration/borrowck/borrowck-borrow-mut-object-twice.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-borrow-mut-object-twice.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-closures-two-mut.rs #53407 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-closures-two-mut-fail.rs#L11
    * but also note there is https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-closures-two-mut-fail.rs#L17
    * and yet there is also this version which uses `borrowck=compare`, (a mode which I think we really should retire at this point...) https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-closures-two-mut.rs#L15 
- [x] ui/compile-fail-migration/borrowck/borrowck-closures-unique-imm.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-closures-unique-imm.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-closures-unique.rs #53407 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-closures-unique.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-issue-2657-1.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-issue-2657-1.rs#L11
- [ ] ui/compile-fail-migration/borrowck/borrowck-lend-flow-if.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-lend-flow-if.rs#L11
- [ ] ui/compile-fail-migration/borrowck/borrowck-lend-flow.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-lend-flow.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-loan-blocks-move-cc.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-loan-blocks-move-cc.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-loan-blocks-move.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-loan-blocks-move.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-move-from-subpath-of-borrowed-path.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-move-from-subpath-of-borrowed-path.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-move-mut-base-ptr.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-move-mut-base-ptr.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-object-lifetime.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-object-lifetime.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-overloaded-index-autoderef.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-overloaded-index-autoderef.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-swap-mut-base-ptr.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-swap-mut-base-ptr.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-union-borrow-nested.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-union-borrow-nested.rs#L11
- [x] ui/compile-fail-migration/borrowck/borrowck-uniq-via-lend.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-uniq-via-lend.rs#L11

## regions/

- [ ] ui/compile-fail-migration/regions-assoc-type-in-supertrait-outlives-container.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-assoc-type-in-supertrait-outlives-container.rs#L11
- [ ] ui/compile-fail-migration/regions-bounded-by-trait-requiring-static.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-bounded-by-trait-requiring-static.rs#L11
- [ ] ui/compile-fail-migration/regions-bounded-method-type-parameters.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-bounded-method-type-parameters.rs#L11
- [ ] ui/compile-fail-migration/regions-free-region-ordering-caller.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-free-region-ordering-caller.rs#L11
- [ ] ui/compile-fail-migration/regions-implied-bounds-projection-gap-1.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-implied-bounds-projection-gap-1.rs#L11
- [ ] ui/compile-fail-migration/regions-infer-contravariance-due-to-decl.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-infer-contravariance-due-to-decl.rs#L11
- [ ] ui/compile-fail-migration/regions-infer-covariance-due-to-decl.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-infer-covariance-due-to-decl.rs#L11
- [ ] ui/compile-fail-migration/regions-outlives-projection-container-wc.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-outlives-projection-container-wc.rs#L11
- [ ] ui/compile-fail-migration/regions-outlives-projection-container.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-outlives-projection-container.rs#L11
- [ ] ui/compile-fail-migration/regions-variance-contravariant-use-covariant-in-second-position.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-variance-contravariant-use-covariant-in-second-position.rs#L11
- [ ] ui/compile-fail-migration/regions-variance-contravariant-use-covariant.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-variance-contravariant-use-covariant.rs#L11
- [ ] ui/compile-fail-migration/regions-variance-covariant-use-contravariant.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-variance-covariant-use-contravariant.rs#L11
- [ ] ui/compile-fail-migration/regions-variance-invariant-use-contravariant.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-variance-invariant-use-contravariant.rs#L11
- [ ] ui/compile-fail-migration/regions-variance-invariant-use-covariant.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/regions/regions-variance-invariant-use-covariant.rs#L11

## variance/

- [ ] ui/compile-fail-migration/variance-cell-is-invariant.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/variance/variance-cell-is-invariant.rs#L11
- [ ] ui/compile-fail-migration/variance-contravariant-arg-trait-match.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/variance/variance-contravariant-arg-trait-match.rs#L11
- [ ] ui/compile-fail-migration/variance-contravariant-self-trait-match.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/variance/variance-contravariant-self-trait-match.rs#L11
- [ ] ui/compile-fail-migration/variance-covariant-arg-trait-match.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/variance/variance-covariant-arg-trait-match.rs#L11
- [ ] ui/compile-fail-migration/variance-covariant-self-trait-match.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/variance/variance-covariant-self-trait-match.rs#L11
- [ ] ui/compile-fail-migration/variance-invariant-arg-trait-match.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/variance/variance-invariant-arg-trait-match.rs#L11
- [ ] ui/compile-fail-migration/variance-invariant-self-trait-match.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/variance/variance-invariant-self-trait-match.rs#L11

## other

- [ ] ui/compile-fail-migration/associated-types-subtyping-1.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/associated-types/associated-types-subtyping-1.rs#L11
- [ ] ui/compile-fail-migration/cleanup-rvalue-scopes-cf.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/cleanup-rvalue-scopes-cf.rs#L11
- [ ] ui/compile-fail-migration/closure_promotion.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/closure_promotion.rs#L11
- [x] ui/compile-fail-migration/hashmap-lifetimes.rs #53407 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/hashmap-lifetimes.rs#L11
- [ ] ui/compile-fail-migration/issue-28848.rs https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/issues/issue-28848.rs#L11
- [ ] ui/compile-fail-migration/issue-5500-1.rs#ast https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/issues/issue-5500-1.rs#L11
- [x] ui/compile-fail-migration/method-self-arg-2.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/methods/method-self-arg-2.rs#L11
- [x] ui/compile-fail-migration/mut-cant-alias.rs #53369 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/mut/mut-cant-alias.rs#L11
- [x] ui/compile-fail-migration/unboxed-closure-region.rs #53407 https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/unboxed-closures/unboxed-closure-region.rs#L11
- [x] ui/compile-fail-migration/unboxed-closures-borrow-conflict.rs #53407  https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/unboxed-closures/unboxed-closures-borrow-conflict.rs#L11

**Comment by pnkfelix** [comment]
I finished going through the remainder of the list.

I have distributed the entries into the following buckets:

### Clear test weakness (in the sense of #51025)

For these, the test presumes lexical lifetimes (e.g. it writes `let _w = &stuff;` or `let _w = &mut stuff;` early on and assumes that borrows lasts until the end of the block). To fix this for NLL, we usually need to add one or more uses of some borrowed variable. I've tried to link to the first such occurrence I noticed in the file.

https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrow-tuple-fields.rs#L21
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-borrow-mut-object-twice.rs#L22
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-closures-unique-imm.rs#L22
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-issue-2657-1.rs#L18
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-loan-blocks-move-cc.rs#L23
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-loan-blocks-move.rs#L20
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-move-from-subpath-of-borrowed-path.rs#L20
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-move-mut-base-ptr.rs#L19
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-mut-borrow-of-mut-base-ptr.rs#L20
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-object-lifetime.rs#L22
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-overloaded-index-autoderef.rs#L46
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-swap-mut-base-ptr.rs#L22
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-union-borrow-nested.rs#L29
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-uniq-via-lend.rs#L44
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/methods/method-self-arg-2.rs#L24
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/mut/mut-cant-alias.rs#L18

### Possible test weakness #51025 (e.g., do we need to add some calls of the closures or object methods?)

https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-closures-two-mut-fail.rs#L25
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-closures-unique.rs#L33

https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/hashmap-lifetimes.rs#L17
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/unboxed-closures/unboxed-closure-region.rs#L16
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/unboxed-closures/unboxed-closures-borrow-conflict.rs#L18


### Fixed (at least somewhat) by NLL

For these cases, we should just add a `#[rustc_error]` to the `fn main()` (assuming the existing `fn main()` is trivial; if the `fn main` is not trivial, then rename it and add a trivial `fn main`)

https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-lend-flow-if.rs#L31-41
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/borrowck/borrowck-lend-flow.rs#L31-37

### Have our rvalue lifetime rules changed...?

I skimmed over this test but I cannot immediately tell whether NLL should have changed anything here... I'll need to read it again with a fresh pair of eyes (and perhaps review some of the doc and/or blog posts discussing our rvalue lifetime rules...)

https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/cleanup-rvalue-scopes-cf.rs#L37-45

### Known bug
These cases all fall into some form of https://github.com/rust-lang/rust/issues/47184

I link to the first obvious example of some type ascription that is currently unchecked: 

https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/closure_promotion.rs#L18
https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/issues/issue-28848.rs#L22

Finally, **all** of the tests in `regions/` and all of the tests in `variance/` use type ascription (in arguably advanced ways) as a way to encode the desired (potentially subtle checks).

I think all of these should be fixed when #47184 is fixed, or potentially sooner as subtasks of that bug are completed.



### Already using -Z borrowck=XXX  in a different revision.

In this case, the test is already requesting some form of MIR-borrowck via explicit compiler flags and presumably does not need to go through the `compare-mode=nll` system. *HOWEVER*, there are instances of this in the test suite where the only reason we used the revision system and compiler flags is that those tests *predate* the compare-mode=nll. So we should consider trying to update them anyway, if possible.

https://github.com/rust-lang/rust/blob/a5733050de780ae4d11e3a7af615df792fdf908e/src/test/ui/issues/issue-5500-1.rs#L13

**Comment by pnkfelix** [comment]
There isn't *that* much left to do here in the short term; IMO all of the regions/ and variance/ are effectively blocked on #47184 -- or at least, I do not think we should invest any effort trying to find a different way to encode those tests. All such effort should be spent instead on addressing #47184, and then once that **is** addressed, we should be able to remove the `// ignore-test-compare-nll` on those files with relative ease.

I think once you remove those files and the others blocked by #47184, then there's only maybe nine or tests left to be either investigated or updated.

**Comment by pnkfelix** [comment]
(having said that, I'm unassigning myself since I am not sure I will have time during the remainder of this week to investigate the remaining tests here.)

**Comment by pnkfelix** [comment]
It might be good to figure out some way to fold this issue (#53351) into #52663, or open up a new issue that covers the remaining cases in the two issues.

Also, if we do one of those big reviews, we may also want to wait until #53764 is done.

**Comment by pnkfelix** [comment]
assigning to self so that remember to factor this into smaller issues that can be individually triaged.

**Comment by pnkfelix** [comment]
I've decided we need a third big round of review, which is filed as #54528

While it might make sense to try to resolve this issue (#53351) before attacking #54528, it probably makes just as much sense to focus attention fully on #54528 itself.

So I'm closing this issue in favor of #54528.
